benchmarks/graphics_pipeline/main.cpp

```diff
@@ -783,6 +835,62 @@ void ProjApp::Setup()
     }
 }
 
+void ProjApp::KeyDown(ppx::KeyCode key)
+{
+    mPressedKeys.insert(key);
+}
+
+void ProjApp::KeyUp(ppx::KeyCode key)
+{
+    mPressedKeys.erase(key);
+}
+
+void FreeCamera::Move(MovementDirection dir, float distance)
+{
+    switch (dir) {
+        case MovementDirection::FORWARD:
+            mEyePosition += float3(0, 0, distance);
+            break;
+        case MovementDirection::LEFT:
+            mEyePosition += float3(distance, 0, 0);
+            break;
+        case MovementDirection::RIGHT:
+            mEyePosition += float3(-distance, 0, 0);
+            break;
+        case MovementDirection::BACKWARD:
+            mEyePosition += float3(0, 0, -distance);
+            break;
+    }
+    mTarget = mEyePosition + float3(0, 0, 1);
+}
+
+void ProjApp::ProcessInput()
+{
+    if (mPressedKeys.empty()) {
+        return;
+    }
+
+    float deltaTime = GetPrevFrameTime();
+
+    if (mPressedKeys.count(ppx::KEY_W) > 0) {
+        mCamera.Move(FreeCamera::MovementDirection::FORWARD, kCameraSpeed * deltaTime);
+    }
+
+    if (mPressedKeys.count(ppx::KEY_A) > 0) {
+        mCamera.Move(FreeCamera::MovementDirection::LEFT, kCameraSpeed * deltaTime);
+    }
+
+    if (mPressedKeys.count(ppx::KEY_S) > 0) {
+        mCamera.Move(FreeCamera::MovementDirection::BACKWARD, kCameraSpeed * deltaTime);
+    }
+
+    if (mPressedKeys.count(ppx::KEY_D) > 0) {
+        mCamera.Move(FreeCamera::MovementDirection::RIGHT, kCameraSpeed * deltaTime);
+    }
+
+    UpdateCamera();
```

Comment:
I find it a bit weird that I could call camera.move() and not move the camera until I call mCamera.LookAt.
Maybe mCamera.LookAt() should be called from inside camera.move(), so it's a real "move" function, not a "prepareMove".